zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentReconstructProcessor.java

```diff
@@ -42,7 +53,134 @@ public DeploymentReconstructProcessor(
   @Override
   public void processRecord(final TypedRecord<DeploymentRecord> record) {
     final var key = keyGenerator.nextKey();
-    // TODO: Only append the event when there are no more deployments to reconstruct
-    stateWriter.appendFollowUpEvent(key, DeploymentIntent.RECONSTRUCTED_ALL, record.getValue());
+    final var resource = findNextResource();
+    if (resource == null) {
+      stateWriter.appendFollowUpEvent(key, DeploymentIntent.RECONSTRUCTED_ALL, record.getValue());
+      return;
+    }
+
+    final DeploymentRecord deploymentRecord;
+    if (resource.deploymentKey() != NO_DEPLOYMENT_KEY) {
+      final var allResourcesOfDeployment = findResourcesWithDeploymentKey(resource.deploymentKey());
+      deploymentRecord =
+          recreateDeploymentForResources(
+              resource.deploymentKey(), resource.tenantId(), allResourcesOfDeployment);
+    } else {
+      deploymentRecord = createNewDeploymentForResource(resource);
+    }
```

Comment:
🙃 Personally, I would flip the statements so there's no negation needed. I always find it reads easier, but it's w/e.